boards/BMS/Src/Io/Io_LTC6813/Io_NewLTC6813.c

```diff
@@ -0,0 +1,321 @@
+    return (uint16_t)(remainder << 1);
+}
+
+void Io_NewLTC6813_InitSpiHandle(SPI_HandleTypeDef *spi_handle)
+{
+    assert(spi_handle != NULL);
+
+    // Initialize the SPI interface to communicate with the LTC6813
+    ltc6813_spi = Io_SharedSpi_Create(
+        spi_handle, SPI2_NSS_GPIO_Port, SPI2_NSS_Pin, SPI_TIMEOUT_MS);
+}
+
+uint16_t Io_NewLTC6813_CalculateRegGroupPec15(uint8_t *data_buffer)
+{
+    return Io_NewLTC6813_CalculatePec15(data_buffer, NUM_OF_BYTES_IN_REG_GROUP);
+}
+
+void Io_NewLTC6813_PackPec15(uint8_t *tx_data, uint8_t num_bytes)
+{
+    const uint16_t cfg_reg_a_pec15 =
+        Io_NewLTC6813_CalculatePec15(tx_data, num_bytes);
+
+    // Pack the PEC15 byte into tx_data in big endian format
+    *((uint16_t *)(tx_data + num_bytes)) =
+        CHANGE_WORD_ENDIANNESS(cfg_reg_a_pec15);
+}
+
+void Io_NewLTC6813_PrepareCmd(uint16_t *tx_cmd, uint16_t cmd)
+{
+    // Change the command endianness
+    *tx_cmd = CHANGE_WORD_ENDIANNESS(cmd);
+
+    // Compute and pack the PEC15 byte into tx_cmd
+    Io_NewLTC6813_PackPec15((uint8_t *)tx_cmd, sizeof(*tx_cmd));
+}
+
+bool Io_NewLTC6813_SendCommand(uint16_t cmd)
+{
+    uint16_t tx_cmd[NUM_OF_CMD_WORDS] = { 0U };
+    Io_NewLTC6813_PrepareCmd(tx_cmd, cmd);
+
+    return Io_SharedSpi_Transmit(
+        ltc6813_spi, (uint8_t *)tx_cmd, TOTAL_NUM_CMD_BYTES);
+}
+
+bool Io_NewLTC6813_PollAdcConversions(void)
+{
+    uint8_t num_attempts = 0U;
+    uint8_t rx_data      = ADC_CONV_INCOMPLETE;
+
+    // Prepare command to get the status of ADC conversions
+    uint16_t tx_cmd[NUM_OF_CMD_WORDS] = { 0 };
+    Io_NewLTC6813_PrepareCmd(tx_cmd, PLADC);
+
+    // All chips on the daisy chain have finished converting cell voltages when
+    // data read back != 0xFF.
+    while (rx_data == ADC_CONV_INCOMPLETE)
+    {
+        const bool is_status_ok = Io_SharedSpi_TransmitAndReceive(
+            ltc6813_spi, (uint8_t *)tx_cmd, TOTAL_NUM_CMD_BYTES, &rx_data,
+            PLADC_RX_SIZE);
+
+        if (!is_status_ok || (num_attempts >= MAX_NUM_ADC_COMPLETE_CHECKS))
+        {
+            return false;
+        }
+
+        num_attempts++;
+    }
+
+    return true;
+}
+
+bool Io_NewLTC6813_WriteConfigurationRegisters(void)
+{
+    for (uint8_t curr_cfg_reg = 0U; curr_cfg_reg < NUM_OF_CFG_REGISTERS;
+         curr_cfg_reg++)
+    {
+        // Command used to write to a configuration register
+        uint16_t tx_cmd[NUM_OF_CMD_WORDS] = { 0U };
+
+        // Array containing bytes to write to the configuration register
+        uint8_t tx_cfg[NUM_OF_ACCUMULATOR_SEGMENTS][NUM_REG_GROUP_BYTES] = {
+            0U
+        };
+
+        // Prepare command to start writing to the configuration
+        // register
+        Io_NewLTC6813_PrepareCmd(tx_cmd, cfg_reg_cmds[curr_cfg_reg]);
+
+        // Prepare default configurations into tx_cfg for each segment
+        memcpy(
+            &tx_cfg[curr_segment], default_cfg_reg[curr_cfg_reg],
+            NUM_REG_GROUP_BYTES);
```

Comment:
I can't find the definition of `curr_segment`. Were you intending on iterating through each segment?

Reply:
yeah this was a mistake on my part, will fix